### ISSUE TITLE
Reachable code in loop marked unreachable

### ISSUE THREAD

**Issue by PeterJCLaw** [python/mypy#8721]
* Are you reporting a bug, or opening a feature request?: bug
* What is the actual behavior/output?

Given the following (cut down) loop code, mypy identifies incorrectly that the print statement cannot be reached.
``` python
last = None
reveal_type(last)  # Revealed type is 'None'

for val in range(10):
    reveal_type(last)  # Revealed type is 'None'

    if last is not None and last < 2:
        print("apparently unreachable")
        reveal_type(last)  # no output

    reveal_type(last)  # Revealed type is 'None'
    last = val
    reveal_type(last)  # Revealed type is 'builtins.int*'
```

My original code also used the value of `last` within the conditional; that was also warned about being not evaluated.

Notably, running `mypy --no-warn-unreachable` still doesn't give any output for the `reveal_type` in the `if` branch.

Happily this issue can be worked around by adding an explicit `Optional[int]` annotation to `last`.

* What is the behavior/output you expect?

I think `mypy` should consider `last` to be `Optional[int]` in the early part of the loop, rather than just being `None`. That would fix the unreachable warning as well as potentially being more correct.

Within the `if last is not None` branch, `last` should be of type `int`.

* What are the versions of mypy and Python you are using? mypy 0.770, Python 3.5.2
  Do you see the same issue after installing mypy from Git master? yes (tested at 125728a547f80ede4ecedc0ff7f890407d309caf)
* What are the mypy flags you are using? `--warn-unreachable`

**Comment by JukkaL** [comment]
This seems pretty bad, as it can result in false negatives. Mypy should give an error when reading the value of `last` before it has a non-`None` type, at least within a loop.

**Comment by davidhalter** [comment]
False positives are also possible without partials, see for example:

https://mypy-play.net/?mypy=latest&python=3.12&flags=warn-unreachable&gist=10808f2ddaa8e35da240e50b068104f3

```
from typing import *                                                            
foo: Union[int, str]                                                            
foo = ""                                                                        
                                                                                
while bool():                                                                   
    reveal_type(foo)                                                            
    if isinstance(foo, str):                                                    
        foo = 1                                                                 
        continue                                                                
    else:                                                                       
        foo = ""                                                                                 
        continue
```
